sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java

```diff
@@ -164,21 +168,20 @@ public CountingSourceReader(int startingPoint) {
 
     @Override
     public boolean start() {
-      return true;
+      return advance();
     }
 
     @Override
     public boolean advance() {
-      if (current < numMessagesPerShard - 1) {
-        // If testing dedup, occasionally insert a duplicate value;
-        if (dedup && ThreadLocalRandom.current().nextInt(5) == 0) {
-          return true;
-        }
-        current++;
-        return true;
-      } else {
+      if (current >= numMessagesPerShard) {
         return false;
       }
+      // If testing dedup, occasionally insert a duplicate value;
+      if (current >= 0 && dedup && ThreadLocalRandom.current().nextInt(5) == 0) {
```

Comment:
Won't this produce duplicates before the checkpoint mark?

So we will output 0, then 1, then checkpoint that we just output 1, and then we will potentially output 1 again after restoring from checkpoint?

It seems as though restoring from checkpoint should guarantee that the previously emit values won't be emit again.